src/pokemon.c

```diff
@@ -0,0 +1,5567 @@
+    monSpDefIV = GetMonData(mon, MON_DATA_SPDEF_IV, 0);
+    monSpDefEV = GetMonData(mon, MON_DATA_SPDEF_EV, 0);
+    monForm = GetMonData(mon, MON_DATA_FORM, 0);
+    monSpecies = GetMonData(mon, MON_DATA_SPECIES, 0);
+    monPersonalData = Heap_AllocFromHeap(0, sizeof(PokemonPersonalData));
+
+    LoadMonFormPersonalData(monSpecies, monForm, monPersonalData);
+
+    if (monSpecies == SPECIES_SHEDINJA) {
+        newMaxHp = 1;
+    } else {
+        newMaxHp = ((2 * monPersonalData->baseHp + monHpIV + monHpEV / 4) * monLevel / 100 + monLevel + 10);
+    }
+
+    SetMonData(mon, MON_DATA_MAX_HP, (u8 *)&newMaxHp);
+
+    newAtk = ((2 * monPersonalData->baseAtk + monAtkIV + monAtkEV / 4) * monLevel / 100 + 5);
+    newAtk = GetNatureAdjustedStatValue(GetMonNature(mon), newAtk, 0x1);
+
+    SetMonData(mon, MON_DATA_ATK, (u8 *)&newAtk);
+
+    newDef = ((2 * monPersonalData->baseDef + monDefIV + monDefEV / 4) * monLevel / 100 + 5);
+    newDef = GetNatureAdjustedStatValue(GetMonNature(mon), newDef, 0x2);
+
+    SetMonData(mon, MON_DATA_DEF, (u8 *)&newDef);
+
+    newSpeed = ((2 * monPersonalData->baseSpeed + monSpeedIV + monSpeedEV / 4) * monLevel / 100 + 5);
+    newSpeed = GetNatureAdjustedStatValue(GetMonNature(mon), newSpeed, 0x3);
+
+    SetMonData(mon, MON_DATA_SPEED, (u8 *)&newSpeed);
+
+    newSpAtk = ((2 * monPersonalData->baseSpAtk + monSpAtkIV + monSpAtkEV / 4) * monLevel / 100 + 5);
+    newSpAtk = GetNatureAdjustedStatValue(GetMonNature(mon), newSpAtk, 0x4);
+
+    SetMonData(mon, MON_DATA_SP_ATK, (u8 *)&newSpAtk);
+
+    newSpDef = ((2 * monPersonalData->baseSpDef + monSpDefIV + monSpDefEV / 4) * monLevel / 100 + 5);
+    newSpDef = GetNatureAdjustedStatValue(GetMonNature(mon), newSpDef, 0x5);
+
+    SetMonData(mon, MON_DATA_SP_DEF, (u8 *)&newSpDef);
+    Heap_FreeToHeap(monPersonalData);
+
+    if ((monCurrentHp == 0) && (monMaxHp != 0)) {
+        (void)0;
+    } else {
+        if (monSpecies == SPECIES_SHEDINJA) {
+            monCurrentHp = 1;
+        } else if (monCurrentHp == 0) {
+            monCurrentHp = newMaxHp;
+        } else {
+            monCurrentHp += (newMaxHp - monMaxHp);
+        }
+    }
+
+    if (monCurrentHp) {
+        SetMonData(mon, MON_DATA_CURRENT_HP, (u8 *)&monCurrentHp);
+    }
+
+    EncryptMon(mon, reencrypt);
+}
+
+u32 GetMonData (Pokemon * mon, int monDataAttribute, void * dest)
+{
+    u32 result;
+    u16 checksum;
+
+    if (mon->box.partyDecrypted == 0) {
+        DecryptMonData(&mon->party, sizeof(PartyPokemon), mon->box.personality);
+        DecryptMonData(&mon->box.dataBlocks, sizeof(PokemonDataBlockA) * 4, mon->box.checksum);
+
+        checksum = GetMonDataChecksum(&mon->box.dataBlocks, sizeof(PokemonDataBlockA) * 4);
+
+        if (checksum != mon->box.checksum) {
+            GF_ASSERT((checksum == mon->box.checksum));
+            mon->box.invalidData = 1;
+        }
+    }
+
+    result = GetMonDataAttribute(mon, monDataAttribute, dest);
```

Comment:
If we keep the name `GetMonData`, then `GetMonDataAttribute` should be `GetMonDataInternal`, and same for `SetMonData`